MM-13537 Improve the UX of the custom theme text box and color selectors

niklabh opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
Improve the UX of the custom theme text box and color selectors
- design changes for color seletors
- color picker hex value input enabled
- only valid hex values allowed
- added copy theme button
- added copy theme success alert badge
![Screenshot from 2019-09-16 00-08-41](https://user-images.githubusercontent.com/874046/64928291-5b0d7700-d816-11e9-8841-7e60a5a82ff3.png)
![Screenshot from 2019-09-16 00-08-53](https://user-images.githubusercontent.com/874046/64928293-5fd22b00-d816-11e9-9c5f-dc06b41f42c2.png)
![Screenshot from 2019-09-16 00-09-17](https://user-images.githubusercontent.com/874046/64928294-6365b200-d816-11e9-94e7-b03998a59645.png)
![Screenshot from 2019-09-16 00-09-26](https://user-images.githubusercontent.com/874046/64928296-652f7580-d816-11e9-8978-318f1caf4509.png)

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://github.com/mattermost/mattermost-server/issues/11951

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

hanzei commented:
Test server is now working :rocket:

esethna commented:
@andrewbrown00 can you take a first pass review at this when you have a chance?

andrewbrown00 commented:
@esethna while I was testing this I noticed the Undo/Redo stack doesn’t respect the HEX value changes. If a user changes a value and uses Edit > Undo or keyboard shortcuts, it should switch back to the last value or values based on the number of changes tracked in the history stack.

Do you want to add this improvement as part of this ticket if it's a major change, or we can create a separate ticket and handle it as a future improvement?

niklabh commented:
Thanks will fix.

andrewbrown00 commented:
@niklabh thanks, let me know if you have any questions?

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

niklabh commented:
I dont know how to do

Don’t select the # as part of the string, even using keyboard shortcuts like CMD+A (Mac) or CTRL+A (Windows). It creates a bug that doesn’t allow the use to input a new value, instead, it places the text cursor at the end of the string.

Pasting is working fine. On complete select and partial select. It only enforces that first character should be hash.

mattermod commented:
Test server creation failed. See the logs for more information.

andrewbrown00 commented:
@niklabh here's the behavior I'm seeing:

1. Select the entire string
2. Start typing or paste a copied HEX value
3. I can't type or paste and the cursor jumps to the end of the string
![cursor-jumps-to-the-end](https://user-images.githubusercontent.com/48559658/65060008-146a6a80-d945-11e9-9d77-0de0a3173491.gif)

If you only select the HEX value and not the # then it works as expected.

niklabh commented:
will fix this

esethna commented:
@andrewbrown00 can you help create a separate ticket for the Undo/Redo stack for this?

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

niklabh commented:
> @niklabh here's the behavior I'm seeing:
> 
> 1. Select the entire string
> 2. Start typing or paste a copied HEX value
> 3. I can't type or paste and the cursor jumps to the end of the string
>    ![cursor-jumps-to-the-end](https://user-images.githubusercontent.com/48559658/65060008-146a6a80-d945-11e9-9d77-0de0a3173491.gif)
> 
> If you only select the HEX value and not the # then it works as expected.

check now

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

mattermod commented:
Test server creation failed. See the logs for more information.

andrewbrown00 commented:
@niklabh it works as long the # is not selected. If the # get selected then the cursor still jumps to the end of the string.

niklabh commented:
I think test server creation failed after latest commit. Can some one please run Setup Cloud Test Server for this again.

niklabh commented:
> @niklabh it works as long the # is not selected. If the # get selected then the cursor still jumps to the end of the string.

This is fixed. Create test server with latest commit to check.

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

mattermod commented:
Test server creation failed. See the logs for more information.

niklabh commented:
> Test server creation failed. See the logs for more information.

Can somebody fix this

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

niklabh commented:
> @niklabh it works as long the # is not selected. If the # get selected then the cursor still jumps to the end of the string.

Its Working now.

niklabh commented:
@andrewbrown00 have you checked

andrewbrown00 commented:
@niklabh sorry, was out of the office yesterday. Things are looking good here 😄 with a few items to resolve.

**Swatch Container**
- [ ] Swatch container is taller than the input
   ![image](https://user-images.githubusercontent.com/48559658/65342002-c9985f00-db9f-11e9-91e7-83fc8c50ba5c.png)
- [ ] Swatch isn't vertically aligned in the container, it has more space on the bottom
   ![image](https://user-images.githubusercontent.com/48559658/65342066-ec2a7800-db9f-11e9-9ddb-65513400900e.png)
- [ ] Swatch container border should match the input border
   ![image](https://user-images.githubusercontent.com/48559658/65331950-bed3cf00-db8b-11e9-8b5d-6f9b67b726a3.png)

**Input Behavior**
- [ ] Clicking in the text input should select the entire HEX value, it currently sets only sets the text cursor at the end of the string. **Tabbing into an input does select the string**
   ![image](https://user-images.githubusercontent.com/48559658/65332215-3bff4400-db8c-11e9-8752-cc0d85525353.png)
- [ ] **We still need some work on the #**
   - [ ] I know you previously mentioned "you can't prevent the # from being selected," but maybe see if another dev can help out to try to solve this. It would add value to the experience because it feels off when the # gets selected which tells the user it will get replaced but it doesn't. The # can be selected and also deleted which feels odd. Can the # be placed in a separate div that is **not selectable**?
   ![image](https://user-images.githubusercontent.com/48559658/65342527-ef723380-dba0-11e9-8a97-854cdcf50495.png)
- [ ] Shorthand HEX values should be converted to 6 digit HEX automatically after input users leaves the input
   ![image](https://user-images.githubusercontent.com/48559658/65341944-a2419200-db9f-11e9-9417-c18f16edad8e.png)
- [ ] Tabbing through the inputs should **first** set focus to the input only, **then** set focus to the entire control and open the swatch selector.  Tabbing for a **third** time should jump to the first tab position on the next input and close the previous controls swatch selector.
   
   **Observed the 'first state' is the only state after tabbing**
   ![image](https://user-images.githubusercontent.com/48559658/65343105-37458a80-dba2-11e9-8cbe-5930260a83eb.png)

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

niklabh commented:
> @niklabh sorry, was out of the office yesterday. Things are looking good here  with a few items to resolve.
> 
> **Swatch Container**
> 
> * [ ]  Swatch container is taller than the input
>   ![image](https://user-images.githubusercontent.com/48559658/65342002-c9985f00-db9f-11e9-91e7-83fc8c50ba5c.png)
> * [ ]  Swatch isn't vertically aligned in the container, it has more space on the bottom
>   ![image](https://user-images.githubusercontent.com/48559658/65342066-ec2a7800-db9f-11e9-9ddb-65513400900e.png)
> * [ ]  Swatch container border should match the input border
>   ![image](https://user-images.githubusercontent.com/48559658/65331950-bed3cf00-db8b-11e9-8b5d-6f9b67b726a3.png)
> 
> **Input Behavior**
> 
> * [ ]  Clicking in the text input should select the entire HEX value, it currently sets only sets the text cursor at the end of the string. **Tabbing into an input does select the string**
>   ![image](https://user-images.githubusercontent.com/48559658/65332215-3bff4400-db8c-11e9-8752-cc0d85525353.png)
> * [ ]  **We still need some work on the #**
>   
>   * [ ]  I know you previously mentioned "you can't prevent the # from being selected," but maybe see if another dev can help out to try to solve this. It would add value to the experience because it feels off when the # gets selected which tells the user it will get replaced but it doesn't. The # can be selected and also deleted which feels odd. Can the # be placed in a separate div that is **not selectable**?
>     ![image](https://user-images.githubusercontent.com/48559658/65342527-ef723380-dba0-11e9-8a97-854cdcf50495.png)
> * [ ]  Shorthand HEX values should be converted to 6 digit HEX automatically after input users leaves the input
>   ![image](https://user-images.githubusercontent.com/48559658/65341944-a2419200-db9f-11e9-9417-c18f16edad8e.png)
> * [ ]  Tabbing through the inputs should **first** set focus to the input only, **then** set focus to the entire control and open the swatch selector.  Tabbing for a **third** time should jump to the first tab position on the next input and close the previous controls swatch selector.
>   **Observed the 'first state' is the only state after tabbing**
>   ![image](https://user-images.githubusercontent.com/48559658/65343105-37458a80-dba2-11e9-8cbe-5930260a83eb.png)

- clicking on input is now selecting hex value. # is not selected
- Shorthand HEX values is converting to 6 digit HEX automatically after input users leaves the input 
- user cannot delete # now.

rest of the fixes can you ask some dev. I am going on a vacation for for days.

mattermod commented:
Test server creation failed. See the logs for more information.

esethna commented:
Thanks @niklabh for your work on this so far, we can wait until your return to continue reviews on this PR. Looking great!

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

niklabh commented:
> > @niklabh sorry, was out of the office yesterday. Things are looking good here  with a few items to resolve.
> > **Swatch Container**
> > 
> > * [x]   Swatch container is taller than the input
> >   ![image](https://user-images.githubusercontent.com/48559658/65342002-c9985f00-db9f-11e9-91e7-83fc8c50ba5c.png)
> > * [x]   Swatch isn't vertically aligned in the container, it has more space on the bottom
> >   ![image](https://user-images.githubusercontent.com/48559658/65342066-ec2a7800-db9f-11e9-9ddb-65513400900e.png)
> > * [x]   Swatch container border should match the input border
> >   ![image](https://user-images.githubusercontent.com/48559658/65331950-bed3cf00-db8b-11e9-8b5d-6f9b67b726a3.png)
> > 
> > **Input Behavior**
> > 
> > * [x]   Clicking in the text input should select the entire HEX value, it currently sets only sets the text cursor at the end of the string. **Tabbing into an input does select the string**
> >   ![image](https://user-images.githubusercontent.com/48559658/65332215-3bff4400-db8c-11e9-8752-cc0d85525353.png)
> > * [ ]   **We still need some work on the #**
> >   
> >   * [ ]   I know you previously mentioned "you can't prevent the # from being selected," but maybe see if another dev can help out to try to solve this. It would add value to the experience because it feels off when the # gets selected which tells the user it will get replaced but it doesn't. The # can be selected and also deleted which feels odd. Can the # be placed in a separate div that is **not selectable**?
> >     ![image](https://user-images.githubusercontent.com/48559658/65342527-ef723380-dba0-11e9-8a97-854cdcf50495.png)
> > * [x]   Shorthand HEX values should be converted to 6 digit HEX automatically after input users leaves the input
> >   ![image](https://user-images.githubusercontent.com/48559658/65341944-a2419200-db9f-11e9-9417-c18f16edad8e.png)
> > * []   Tabbing through the inputs should **first** set focus to the input only, **then** set focus to the entire control and open the swatch selector.  Tabbing for a **third** time should jump to the first tab position on the next input and close the previous controls swatch selector.
> >   **Observed the 'first state' is the only state after tabbing**
> >   ![image](https://user-images.githubusercontent.com/48559658/65343105-37458a80-dba2-11e9-8cbe-5930260a83eb.png)
> 
Checked ones i have fixed. Rest 2 issues are too difficult for me to fix.
Rest

mattermod commented:
Test server creation failed. See the logs for more information.

hanzei commented:
@niklabh Your tests are failing: https://circleci.com/build-insights/gh/mattermost/mattermost-webapp/pull%252F3672

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

niklabh commented:
updated snapshots.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

andrewbrown00 commented:
@niklabh thanks for all the work on this, it looks great 👍. We'll pull in another resource to knock out the outstanding items.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

esethna commented:
@andrewbrown00 do you know what are next steps on this ticket?

andrewbrown00 commented:
@esethna it has a few outstanding items that need to be resolved. I chatted with @hmhealey and he is going to make the changes but we need to get the ticket into the upcoming sprint.

Here's the Jira ticket with the comments: https://mattermost.atlassian.net/browse/MM-13537

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `2183097fe6a82208b7f0859dd3b2af7db6263e25`.

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

hmhealey commented:
This got missed again in sprint planning. I'll make sure it gets planned for next week if I don't have a chance to get to it sooner

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `e732adf8cb4f8494b4a63a8312a56c5004858d37`.

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Test server creation failed. See the logs for more information.

andrewbrown00 commented:
This is on hold while @hmhealey finishes up the Mark as Unread feature.

esethna commented:
Removing from the v5.18 milestone

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `42f5a088487c237b690c5b34ad64a7395fe5249a`.

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

mattermod commented:
Test server destroyed

hanzei commented:
Hi @niklabh,

Will do have the time to continue working on this PR?

Please note that we will close this PR after another 10 days of inactivity. If you don't object, other contributors might complete this PR for you.

niklabh commented:
> @esethna it has a few outstanding items that need to be resolved. I chatted with @hmhealey and he is going to make the changes but we need to get the ticket into the upcoming sprint.
> 
> Here's the Jira ticket with the comments: https://mattermost.atlassian.net/browse/MM-13537

@hanzei As per previous conversation with @andrewbrown00 @hmhealey was taking care of the unresolved issue. Let me know what is needed to be done from my side. I would be happy to assist.

hanzei commented:
It seams like to completely misunderstood the conversation, sorry @niklabh. 

@bradjcoughlin per the JIRA ticket you will help @niklabh with finishing this PR- Could you please share a status update?

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

hmhealey commented:
I don't believe Brad will be back until Monday, so I assume he can give an update then

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

jasonblais commented:
Hey @bradjcoughlin, friendly reminder to help share a status or next steps on this PR? 

https://github.com/mattermost/mattermost-webapp/pull/3672#issuecomment-568156648

bradjcoughlin commented:
/update-branch

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

bradjcoughlin commented:
The decision was made by @andrewbrown00 and @adamjclarkson to go ahead and ship what is in this PR, creating another ticket for the additional accessibility improvement. Tagging for dev review.

niklabh commented:
Do i need to merge master into this?

bradjcoughlin commented:
@niklabh nope, I did that with the '/update-branch' action above. Just waiting on a second dev review and then we’ll merge. Thanks again for all your work on this!

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `c979ac45c6a439dfed4289ed03f58ec6655456c8`.

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

Willyfrog commented on components/user_settings/display/user_settings_theme/custom_theme_chooser.jsx:
```diff
@@ -376,18 +393,34 @@ export default class CustomThemeChooser extends React.Component {
+                        className='btn btn-link copy-theme-button'
+                        onClick={this.copyTheme}
+                    >
+                        {'Copy Theme Colors'}
```
I think we should use FormattedMessage here, otherwise it won't be translated

niklabh replied:
done

Willyfrog commented on components/user_settings/display/user_settings_theme/custom_theme_chooser.jsx:
```diff
@@ -376,18 +393,34 @@ export default class CustomThemeChooser extends React.Component {
+                        role='alert'
+                        style={{display: 'none'}}
+                    >
+                        {'✔ Copied'}
```
use FormattedMessage

niklabh replied:
done

bradjcoughlin commented on components/color_input.tsx:
```diff
@@ -4,6 +4,8 @@
 import React from 'react';
 import {ChromePicker, ColorResult} from 'react-color';
 
+const hexRegex = /^#[0-9a-fA-F]*$/i;
```
@niklabh If you don't have time to make the following changes, let me know. I'd like to get this feature merged, so I'm happy to help out.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `dfd0261ec9f08d6f0a6eb533471b14b5235ea962`.

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

niklabh commented on components/color_input.tsx:
```diff
@@ -12,23 +14,35 @@ type Props = {
+        this.colorInput = React.createRef();
         this.state = {
             isOpened: false,
+            hex: '#FFFFFF',
```
state.hex is used. changed white to use props.color

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

niklabh commented on components/color_input.tsx:
```diff
@@ -57,22 +78,85 @@ class ColorInput extends React.PureComponent<Props, State> {
+        }
+        if (value.length > 7) {
+            return;
+        }
```
done

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

niklabh commented on components/color_input.tsx:
```diff
@@ -57,22 +78,85 @@ class ColorInput extends React.PureComponent<Props, State> {
+        }
+        this.setState({hex: value});
+        const {onChange: handleChange} = this.props;
+        if (handleChange) {
```
changed to handleColorChange

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

niklabh commented on components/color_input.tsx:
```diff
@@ -57,22 +78,85 @@ class ColorInput extends React.PureComponent<Props, State> {
     };
 
+    private getShortHand = (value: string) => {
+        return value.split('').map((ch, index) => {
```
done

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

niklabh commented on components/color_input.tsx:
```diff
@@ -57,22 +78,85 @@ class ColorInput extends React.PureComponent<Props, State> {
+            }
+            if (value.length === 4) {
+                handleChange(this.getShortHand(value));
+            }
```
done

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Mattermost test server updated with git commit `27fb29530ab8ac665e1a6d9b15341ed4ffa590aa`.

Access here: https://mattermost-webapp-pr-3672.test.mattermost.cloud

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

niklabh commented:
> looking good, seems like you forgot to add the i18n lines for the new formatted messages and it would be good from my perspective.

done

Willyfrog commented on i18n/en.json:
```diff
@@ -3356,8 +3356,10 @@
   "user.settings.custom_theme.centerChannelColor": "Center Channel Text",
   "user.settings.custom_theme.centerChannelTitle": "Center Channel Styles",
+  "user.settings.custom_theme.copied": "✔ Copied",
   "user.settings.custom_theme.codeTheme": "Code Theme",
```
order matters:
```
  "user.settings.custom_theme.codeTheme": "Code Theme",
  "user.settings.custom_theme.copied": "✔ Copied",
```
code should go before copied
